### DISCUSSION TITLE
Local development with Flux: no --git-readonly option?

### DISCUSSION THREAD

**Discussion by ariehm** [fluxcd/flux2#4867] [Q&A] [answered]
# The Setup

Let me apologize up-front for my ignorance.  I'm new to Flux, using it on a small team, and loving it - it works great for syncing our config in a private GitHub repository with a Kubernetes cluster in The Cloud.

But I'd like to be able to do local dev in a representative way, so I ran `flux bootstrap` to wire up a local minikube cluster and point it at `clusters/local/` in our private Flux repo.

So far, so good.  I can reuse a lot of the same configurations in that `clusters/local/` folder that I'm using in our `clusters/production/` folder, and so quickly spin up a representative environment locally.  Great!

# The Problem

The problem I ran into is two-fold:
* There are other devs on the team, and ideally we'd all be able to spin up local environments easily, pointed at the same `clusters/local/` folder, so we all have the same setup locally.
* It be nice to be able to point my local minikube instance at a branch, so I don't have to go through approved-PRs just to see if my local changes worked locally.

In either case, any new `flux bootstrap` commands seem to (with an admin GITHUB_TOKEN) push changes to whatever branch you point the command at, and that causes conflicts.  Plus, we'd like to be able to set this up on every dev's machine, but we don't want every dev on the team to have admin powers - especially against the same repo that manages our production configurations.

# --git-readonly To The Rescue?

I started looking for a "flux readonly" option, and I came across this Discussion: https://github.com/fluxcd/flux2/discussions/1819

(among other references: seems like this was a thing in V1 - https://github.com/fluxcd/flux/issues/1139)

The reason I'm posting this on a new Discussion is that my use case seems slightly different from that 1819, and I didn't want to post this as an "answer" on that thread.

It does seem like `flux bootstrap --git-readonly` would solve my problems in a magical way, whereas `flux install` (and then do stuff to bridge the gap, manually, every time I switch branches or spin up a new dev on the team) seems rough.

# Bridging The `flux install` Gap?

If I just run `flux install`, that installs flux resources fine, but I'm not sure how to get from `flux install` to "I now have a local minikube cluster with flux pointed at a certain path on a branch in a git repo".  Are there docs on how to bridge this gap?

All the docs I could find on how to point Flux at a Github repo suggest using `flux bootstrap`, which, according to the official v2 docs (https://fluxcd.io/flux/installation/bootstrap/github/) seems to require an admin token on the repo.

Is it really as simple as running `flux install` and then `kubectl apply -f <A FILE WITH A GitRepository CRD>` (secrets probably also need to get in there to, for accessing the private repo).

These docs (https://fluxcd.io/flux/components/source/gitrepositories/) are cool, but don't seem to mention if that's the right way to do what I'm trying to do.

# Am I Doing It Wrong?

Is there an easier way to both use flux for managing production environments *and* spinning up representative local development environments, across branches and multiple team members, that doesn't require read-write GitHub access?

Again, my apologies for both my ignorance and long-windedness.  Can anyone point me at some useful docs, or suggest a different approach that would work?

**Comment by kingdonb** [comment]
There are a couple of issues with your idea, but they can be worked around. The short answer, Flux bootstrap is readonly except for the Git Deploy Key, which can't be avoided unless your sources are public. But you can create deploy keys manually with `flux create secret git`

Take a look at this "Identical environments" here

https://github.com/fluxcd/flux2-kustomize-helm-example?tab=readme-ov-file#identical-environments

If you have followed the docs this far then you have a directory structure like in flux2-kustomize-helm-example -- most of the work does not go into `clusters/production` or `clusters/local` - flux resources are defined there, kustomization overlays can be defined there to avoid duplicating flux resource definitions that should be reused without modification (or even with some patches) but the bulk of your definitions will go into `apps` and `infrastructure` so the definitions can be used by multiple clusters without copying.

The problem that remains is you cannot bootstrap twice from `production-clone` - but you said you want multiple devs to be able to work on this at once.

Two ways to solve this: one, add `production-clone2` which is just another copy, and repeat ad-infinitum (or eg. `production-clone-bob` if it's meant to be one environment per developer)

Why: If you have two devs working on different stuff, there's no reason they should share the same definition of `production-clone` - what if one of them wants to add patches, will they consult the other devs first? Problems caused by cross-talking between their experiments? They should have separate definitions.

But maybe not. Conway's law, maybe your situation is different and you know perfectly well what you want. But when you bootstrap the same path a second time, as you saw, Flux errors out warning you that an installation of Flux is already bootstrapped on that path.

There are a couple of reasons why that won't work, but basically the deploy keys added to bootstrap repos are keyed by context and path. So this is intentionally blocked by Flux bootstrap CLI. Adding another bootstrap accidentally with the same context and path is prevented through this keying. But what if you really want identical environments (and you know you really don't want it to diverge?)

Well, you can use this same structure but do not bootstrap. Make `production-clone` but instead of bootstrapping, take a copy of `flux-system` and edit the `spec.path` manually. Commit it to the repo as it would be done if you ran bootstrap. Apply the copy to the cluster with `kubectl apply -k flux-system`. If your cluster repo is private as it probably should be, you have the additional step of handling a SSH deploy key. Use `flux create secret git` for this to be sure you don't make any copy-paste errors creating the secret. Apply the `flux-system` directory to the cluster by hand, and take care that you use the correct path because there are no guard rails.

Be aware that when you create a PAT for bootstrapping or create deploy keys manually, they can have their own lifecycle that is bound to your user - if you create deploy keys by hand or store a PAT to access the repo, this behavior is different from case to case. If your cluster repo is public (probably shouldn't be) you can edit that copy of `gotk-sync.yaml` and use the public cluster repo URL. Then remove the secretRef and apply it directly to the cluster with the whole `flux-system` directory.

This public repo approach might still work well if you are using a Git URL behind the firewall, but usually you should authenticate.

You can do this as many times as you want on different clusters, they will never conflict.

**Reply by kingdonb** [reply, marked as answer]
Also, rereading your two-part question:

> It be nice to be able to point my local minikube instance at a branch, so I don't have to go through approved-PRs just to see if my local changes worked locally.

We don't really recommend branches as a strategy for concurrent environments, because there is no clean answer for what do you do when a change in one environment destined for another environment merges back?

You have to be careful not to merge the `spec.ref.branch` change back with the PR and its other changes. This is complicated. Kargo is one solution for that type of issue, but it may not be relevant here, only to say that the delivery problem (of shipping changes from one environment to downstream environments) is complex and deserves a separate treatment.

You can use github forks (or gitlab forks) so you don't have to give each dev admin access to the central repo, and they can bootstrap their own minikubes from the fork. I don't think you have to abandon the `flux bootstrap github` CLI then, since every dev can have their own fork of the platform repo, and they will have admin access to their own fork (but not shared with others) they are also free to diverge from one another and will not interfere with each other when they create patches to the local instance.

The devs can keep their test changes in the `main` branch of their fork and bootstrap will (or they can) change the clone git url to point at their fork. This will cause it to diverge a little bit from upstream, (either use the full read-only method to avoid that or ensure that devs are prepared to learn rebasing.)

This fork-based approach to dev environments is a form of "hard multitenancy".

**Reply by ariehm** [reply]
Thank you for your prompt and thorough reply!  There are a few different problems I'm trying to solve with Flux, and maybe the answer for one will be different than the other, but either way the information you provided is most illuminating.

For clarity's sake, I'd like to point out two different problems (general problems, not Flux problems) we are facing:

* A few people on our team are working on infrastructure-related things, and that involves working with Flux
* More people on our team are working on application development locally - applications that we then deploy to staging and production environments using Flux.

## Local Prototyping for the Infrastructure Team

It sounds like if the (handful of) folks on our team that are working on infrastructure need to spin up local Kubernetes things to prototype stuff using Flux, we can fork the main repo, get the prototypes working, and then port those changes back to the main repo via the usual mechanisms.  This should work great - thanks for pointing this out as an option!

## Local Application Development

A separate problem is how to ensure all our application devs have a consistent development environment applied across all their machines, and it sounds like Flux is not the right tool for that job:
* every developer having commit access to our flux repo (that controls our production environment) sounds like a bad idea, 
* keeping a dozen different folders in the flux repo in sync sounds like a bad way to go
* every developer having their own fork of the Flux repo sounds like even more of a synchronization nightmare 

I'm trying to avoid the "it works on *my* box" problem, because of drift between repositories and differences in local environments that happen outside of Flux (manually applying Kubernetes manifests, for example).

I was hoping to deploy a base configuration to everyone's local clusters with Flux that did *not* include our applications (but *did* include all of our standard infrastructure and tooling, Istio, mesh configurations, observability tools, etc).  That way everyone could have a standard set of things that I keep up to date with the staging and prod configurations, automatically synced to their boxes without requiring much from them.  But it sounds like Flux is not the way to do that.

## The Point

My apologies again for both my ignorance and long-windedness.  Thanks for answering my question - I think we've got a good way to move forward for the one problem, and I'll look elsewhere for solutions to the local dev environment problem.

Thanks again for your response, and thanks for all the hard work on Flux - it's an amazing tool!

## PS

I've seen this expressed in your answer here and on other posts (from others) that I found when Googling, and I want to clear something up:

> Flux bootstrap is readonly except for the Git Deploy Key

This is probably my fault for not being more clear.  When I say that I want `flux bootstrap --readonly` (if it existed) to be read-only, what I mean by that is that I don't want it to commit anything to the repo at which I'm pointing the command.

When I run `flux boostrap`, as far as I can tell, it makes commits on the repo+branch at which you're pointed (unless it doesn't detect any difference between the manifests in the repo and what it generates locally?).  Those commits bypass the PR process, don't go through approvals, have no chance to be reviewed or rejected.

It's super handy for most use-cases, but could potentially cause all sorts of problems if multiple devs are running `flux bootstrap` on their boxes, pointed at their local clusters and the same repo/branch/path - which, as you made clear earlier, was a bad idea for all sorts of reasons.

I just wanted to clear up what I mean when I say "read-only" - namely , `flux bootstrap --readonly` (if it existed) would not attempt to write commits to the repository.

I hope that makes sense.  Thanks again for your help!

**Reply by kingdonb** [reply]
> Flux bootstrap is readonly except for the Git Deploy Key

This is not really accurate, of course, bootstrap does commit to the repo as well as the deploy key

Many people have issues with bootstrap, starting with "how do you bootstrap in an environment where there are branch protections in place" - you can't, not without some weird branch-change-commit-before-merge hand-off thing.

You might have better luck with the `flux-operator` project that is designed to solve some of the problems that people have with bootstrapping:

https://fluxcd.control-plane.io/operator/flux-bootstrap-migration/

This is a first-class supported Flux operator, it's a bit new, but it could neatly solve your issue. The main issue with bootstrap forks and branch per environment is the big mess of YAML that gets out of sync, and it's not immediately clear on inspection from one version to the next, what's different between them.

The `FluxInstance` is really simplified in terms of YAML and intents. You can do away with the whole `flux-system` directory and make bootstrap much simpler. You don't really need bootstrap anymore, you can keep Flux updated with patch releases and keep everyone's branches and forks straight easily because the diff between two different versions of Flux will be a one-line change.

https://fluxcd.control-plane.io/operator/fluxinstance/

Maybe this helps? It is not GA like Flux Bootstrap, but I have shown this to a couple of people so far and so far it's getting rave reviews.

**Reply by ariehm** [reply]
![Screenshot 2024-07-01 at 5 05 02 PM](https://github.com/fluxcd/flux2/assets/156836989/276f7b5e-0068-4cef-a274-78b20e84d940)

I like it already! :) Thanks again for your help!
